Yoast/wordpress-seo
# Do not load current user too early

## Summary

This PR can be summarized in the following changelog entry:

* Ensures the current user is not initialized before its expected core initialization by deferring all user-related functionality to the 'init' hook or later.

## Relevant technical choices:

* An alternative approach could have been to initialize the entire notification center later, but there are many other parts of the plugin needing it earlier. Therefore it's a better idea to solve the problem internally.
* The user storage is now not touched until the `init` hook. If notifications are being added or removed before, these transactions are then stored in a queue. When the current user's notifications are fetched from the storage, previously existing transactions are then applied (in the same order they were originally supposed to happen).
* There is one additional case in the plugin where I found that the code initialized the current user too early (`WPSEO_Admin::check_php_version()`). Since that method isn't complex and doesn't tie in with other functionality that would be executed before `init`, I ensured this is run on `init` as well instead of directly in the constructor (which is run on `plugins_loaded`, so too early).

## Test instructions

This PR can be tested by following these steps:

* Verify that the same notifications as before are present in your installation.
* On outdated PHP versions, make sure you still see the WHIP notice (as an administrator) like you should have before.
* Add a snippet (preferably in an mu-plugin) where you hook into the `wp_set_current_user` action. In there, check `doing_action( 'plugins_loaded' )` and ensure it evaluates to false. If it does not evaluate to false, inspect the current state of execution further (for example by checking backtrace). I discovered there is a similar bug in WordPress core that was introduced in 4.7 - any call to `load_plugin_textdomain()` will also trigger the user to be set too early (I'll open a core ticket for it shortly). So if that function is responsible for it, ignore it. Only ensure it doesn't happen because of other Yoast SEO functionality.

## Quality assurance


Fixes #9452
